File: src/main/java/com/bsoft/blink_api/controllers/AuthController.java
package com.bsoft.blink_api.controllers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.bsoft.blink_api.models.Payload;
import com.bsoft.blink_api.models.RegistrationForm;
import com.bsoft.blink_api.services.CustomUserService;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;


@RestController
@RequestMapping("/auth")
public class AuthController {
    @Autowired CustomUserService userService;

    @PostMapping("/login")
    public String login(@RequestBody String entity) { 
        return entity;
    }

    @PostMapping("/signin")
    public Payload<?> signin(@RequestBody RegistrationForm form) {
        System.out.println(form.password());
        System.out.println(form.matchingPassword());
        if(form.password().equals(form.matchingPassword())){
            try {
                UserDetails details = userService.register(form);

                return new Payload<UserDetails>(201, details);
            } catch (Exception e) {
                System.out.println(e);
                return new Payload<String>(500, "user registration failed");
            }
        }
        return new Payload<String>(400, "password mismatch");
    }
}

File: src/controllers/auth-controller.ts
import { authRepository, userRepository } from "../repositories";
import { Exception } from "../repositories/models";

namespace authController{
    export const singup = async (data: { fname: string, lname: string, email: string, password: string, matchingPassword: string}) =>{
        if(data.password === data.matchingPassword){
            const user = await userRepository.create(data)
            const auth = await authRepository.create({ userID: user.id, password: data.password })

            return { auth, user };
        }
        throw new Exception(401, "password mismatch")
    }

    export const login = async (data: { email: string, password: string}) =>{
        const user = await userRepository.find(data.email)
        const auth = await authRepository.get(user.id)

        if(auth.password === data.password){
            return { auth, user }
        }
        throw new Exception(401, "incorrect password")
    }
}

export default authController